third/bind/doc/man/getnetent.3
.so /usr/athena/share/tmac/tmac.doc
.\" $Id: getnetent.3,v 1.2 1998-12-18 17:22:51 danw Exp $
.\"
.\"Copyright (c) 1995, 1996 by Internet Software Consortium
.\"
.\"Permission to use, copy, modify, and distribute this software for any
.\"purpose with or without fee is hereby granted, provided that the above
.\"copyright notice and this permission notice appear in all copies.
.\"
.\"THE SOFTWARE IS PROVIDED "AS IS" AND INTERNET SOFTWARE CONSORTIUM DISCLAIMS
.\"ALL WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES
.\"OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL INTERNET SOFTWARE
.\"CONSORTIUM BE LIABLE FOR ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL
.\"DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR
.\"PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS
.\"ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS
.\"SOFTWARE.
.\"
.Dd May 20, 1996
.Dt GETNETENT @LIB_NETWORK_EXT_U@
.Os BSD 4
.Sh NAME
.Nm getnetent ,
.Nm getnetbyaddr ,
.Nm getnetbyname ,
.Nm setnetent ,
.Nm endnetent 
.Nd get networks entry
.Sh SYNOPSIS
.Fd #include <netdb.h>
.Ft struct netent *
.Fn getnetent
.Ft struct netent *
.Fn getnetbyname "char name";
.Ft struct netent *
.Fn getnetbyaddr "unsigned long net" "int type";
.Ft void 
.Fn setnetent "int stayopen";
.Ft void 
.Fn endnetent
.Sh DESCRIPTION
The
.Fn getnetent ,
.Fn getnetbyname ,
and
.Fn getnetbyaddr
subroutines
each return a pointer to an object with the following structure
containing the broken-out fields of a line in the 
.Pa networks 
database.
.Bd -literal -offset indent
struct	netent {
	char	*n_name;	/* official name of net */
	char	**n_aliases;	/* alias list */
	int	n_addrtype;	/* net number type */
	long	n_net;		/* net number */
};
.Ed
.Pp
The members of this structure are:
.Bl -tag -width "n_addrtype" 
.It n_name
The official name of the network.
.It n_aliases
A zero-terminated list of alternate names for the network.
.It n_addrtype
The type of the network number returned: 
.Dv AF_INET .
.It n_net
The network number.  Network numbers are returned in machine byte
order.
.El
.Pp
If the
.Fa stayopen
flag on a 
.Fn setnetent
subroutine is NULL, the
.Pa networks
database is opened.  Otherwise, the
.Fn setnetent
has the effect of rewinding the 
.Pa networks 
database.
The
.Fn endnetent
subroutine may be called to
close the 
.Pa networks 
database when processing is complete.
.Pp
The
.Fn getnetent
subroutine simply reads the next
line while
.Fn getnetbyname
and
.Fn getnetbyaddr
search until a matching
.Fa name
or
.Fa net
number is found
(or until 
.Dv EOF 
is encountered).  The 
.Fa type must be 
.Dv AF_INET .
The
.Fn getnetent
subroutine keeps a pointer in the database, allowing
successive calls to be used to search the entire file.
.Pp
Before a
.Ic while
loop using
.Fn getnetent ,
a call to
.Fn setnetent
must be made 
in order to perform initialization; a call to 
.Fn endnetent
must be used after the loop.  Both
.Fn getnetbyname
and
.Fn getnetbyaddr
make calls to
.Fn setnetent
and
.Fn endnetent .
.Sh FILES
.Pa /etc/networks
.Sh DIAGNOSTICS
Null pointer (0) returned on 
.Dv EOF 
or error.
.Sh SEE ALSO
.Xr networks @FORMAT_EXT@ ,
RFC 1101.
.Sh HISTORY
The 
.Fn "getnetent" , 
.Fn "getnetbyaddr" , 
.Fn "getnetbyname" , 
.Fn "setnetent" , 
and
.Fn "endnetent"
functions appeared in 4.2BSD.
.Sh BUGS
The data space used by these functions is static; if future use requires the
data, it should be copied before any subsequent calls to these functions
overwrite it.  Only Internet network numbers are currently understood.
Expecting network numbers to fit in no more than 32 bits is probably naive.


doc/news
This file is intended to keep track of user-visible changes to the
release, for the construction of the Athena release notes.

Changes from release 8.0 to release 8.1:

	* Athena version numbers are now dotted triplets, e.g. 8.1.0
	  or 8.1.12.  Paths which include all or part of the Athena
	  version number now take the form "pathname-version",
	  e.g. "srvd-8.1" or "dash.menus-8.1"

	* In release 8.0, running "xmkmf" with the default path would
	  generate a Makefile which looked for its include files and
	  libraries in /mit/x11, which could be one of several
	  lockers.  Now it is preferrable to use the vendor X
	  libraries, so this bit of glue is going away.  There will be
	  no /usr/athena/bin/xmkmf any more, nor any symlinks in
	  /usr/athena pointing to the x11 locker.

	  Users can still do something like "add -f x11r5" and then
	  "xmkmf", and things should work, if the x11r5 locker is
	  built for their platform.

	* Athena bitmaps have moved to /usr/athena/include/X11/bitmaps.
	  Previously, they were in /usr/athena/lib/X11/bitmaps before
	  because of the /mit/x11 glue.

	* The symlink /bin/csh -> /bin/athena/tcsh has gone away.  The
	  update to 8.0 will fix /bin/csh references in /etc/passwd
	  and /etc/passwd.local to refer to /bin/athena/tcsh instead.

	* emacs18 has gone away.

	* AFS includes and libraries are installed in
	  /usr/afsws/include and /usr/afsws/lib.

	* wcl includes and libraries are installed in /usr/athena.

	* The basic krb5 headers and libraries are installed in
	  /usr/athena.

	* emacs has been upgraded from 19.30 to 19.34.  There should
	  be no significant user-visible changes.

	* RCS has been upgraded from version 4 to version 5.7.

	* x3270 has been upgraded from version 1.2 to version 3.1.1.3.

	* sysinfo is now available under Irix.

	* /bin/athena/tcsh now has Hesiod support enabled, which means
	  that "~lockername" will work in a pathname in most cases.
	  (It won't work for users with local passwd information who
	  don't have their lockers attached.)

	* For improved ease of debugging, many binaries are installed
	  with symbol tables, and compiled without optimization.
	  Therefore, the contents of /bin/athena and /usr/athena/bin
	  may be somewhat larger, and the binaries may consume more
	  CPU cycles, although it's unlikely that anyone will notice
	  the speed difference.


third/bind/src/lib/irs/getnetent.c
/*
 * Copyright (c) 1996 by Internet Software Consortium.
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND INTERNET SOFTWARE CONSORTIUM DISCLAIMS
 * ALL WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL INTERNET SOFTWARE
 * CONSORTIUM BE LIABLE FOR ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL
 * DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR
 * PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS
 * ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS
 * SOFTWARE.
 */

#if !defined(LINT) && !defined(CODECENTER)
static char rcsid[] = "$Id: getnetent.c,v 1.1.1.1 1998-05-04 22:23:40 ghudson Exp $";
#endif

/* Imports */

#include "port_before.h"

#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/nameser.h>
#include <arpa/inet.h>

#include <ctype.h>
#include <errno.h>
#include <netdb.h>
#include <stdlib.h>
#include <string.h>

#include "port_after.h"

#include "irs.h"
#include "irs_p.h"
#include "irs_data.h"

/* Definitions */

struct pvt {
	struct netent	netent;
	char *		aliases[1];
	char		name[MAXDNAME + 1];
};

/* Forward */

static struct irs_nw *	init(void);
static struct netent *	nw_to_net(struct nwent *);
static void		freepvt(void);
static struct netent *	fakeaddr(const char *, int af);

/* Portability */

#ifndef INADDR_NONE
# define INADDR_NONE 0xffffffff
#endif

/* Public */

struct netent *
getnetent() {
	struct irs_nw *nw = init();
	
	if (!nw)
		return (NULL);
	net_data.nw_last = nw_to_net((*nw->next)(nw));
	return (net_data.nw_last);
}

struct netent *
getnetbyname(const char *name) {
	struct irs_nw *nw = init();
	struct netent *np;
	char **nap;
	
	if (!nw)
		return (NULL);
	if (net_data.nw_stayopen && net_data.nw_last) {
		if (!strcmp(net_data.nw_last->n_name, name))
			return (net_data.nw_last);
		for (nap = net_data.nw_last->n_aliases; nap && *nap; nap++)
			if (!strcmp(name, *nap))
				return (net_data.nw_last);
	}
	if ((np = fakeaddr(name, AF_INET)) != NULL)
		return (np);
	net_data.nw_last = nw_to_net((*nw->byname)(nw, name, AF_INET));
	if (!net_data.nw_stayopen)
		endnetent();
	return (net_data.nw_last);
}

struct netent *
getnetbyaddr(unsigned long net, int type) {
	struct irs_nw *nw = init();
	u_char addr[4];
	int bits;
	
	if (!nw)
		return (NULL);
	if (net_data.nw_stayopen && net_data.nw_last)
		if (type == net_data.nw_last->n_addrtype &&
		    net == net_data.nw_last->n_net)
			return (net_data.nw_last);
	
	addr[3] = (0xFF000000 & net) >> 24;
	addr[2] = (0x00FF0000 & net) >> 16;
	addr[1] = (0x0000FF00 & net) >> 8;
	addr[0] = (0x000000FF & net);
	
	/* Use the old class rules to figure out the network bits. */
	if (addr[3] >= 240)
		bits = 32;
	else if (addr[3] >= 224)
		bits = 4;
	else if (addr[3] >= 192)
		bits = 24;
	else if (addr[3] >= 128)
		bits = 16;
	else
		bits = 8;
	
	net_data.nw_last = nw_to_net((*nw->byaddr)(nw, addr, bits, AF_INET));
	if (!net_data.nw_stayopen)
		endnetent();
	return (net_data.nw_last);
}

void
setnetent(int stayopen) {
	struct irs_nw *nw = init();
	
	if (!nw)
		return;
	freepvt();
	(*nw->rewind)(nw);
	net_data.nw_stayopen = (stayopen != 0);
}

void
endnetent() {
	struct irs_nw *nw = init();

	if (nw != NULL)
		(*nw->minimize)(nw);
}

/* Private */

static struct irs_nw *
init() {
	if (!net_data_init())
		goto error;
	if (!net_data.nw)
		net_data.nw = (*net_data.irs->nw_map)(net_data.irs);
	if (!net_data.nw) {
 error:		
		errno = EIO;
		return (NULL);
	}
	return (net_data.nw);
}

static void
freepvt() {
	if (net_data.nw_data) {
		free(net_data.nw_data);
		net_data.nw_data = NULL;
	}
}

static struct netent *
fakeaddr(const char *name, int af) {
	struct pvt *pvt;
	const char *cp;
	u_long tmp;

	if (af != AF_INET) {
		/* XXX should support IPv6 some day */
		errno = EAFNOSUPPORT;
		h_errno = NETDB_INTERNAL;
		return (NULL);
	}
	if (!isascii(name[0]) || !isdigit(name[0]))
		return (NULL);
	for (cp = name; *cp; ++cp)
		if (!isascii(*cp) || (!isdigit(*cp) && *cp != '.'))
			return (NULL);
	if (*--cp == '.')
		return (NULL);

	/* All-numeric, no dot at the end. */

	tmp = inet_network(name);
	if (tmp == INADDR_NONE) {
		h_errno = HOST_NOT_FOUND;
		return (NULL);
	}

	/* Valid network number specified.
	 * Fake up a netent as if we'd actually
	 * done a lookup.
	 */
	freepvt();
	net_data.nw_data = malloc(sizeof(struct pvt));
	if (!net_data.nw_data) {
		errno = ENOMEM;
		h_errno = NETDB_INTERNAL;
		return (NULL);
	}
	pvt = net_data.nw_data;

	strncpy(pvt->name, name, MAXDNAME);
	pvt->name[MAXDNAME] = '\0';
	pvt->netent.n_name = pvt->name;
	pvt->netent.n_addrtype = AF_INET;
	pvt->netent.n_aliases = pvt->aliases;
	pvt->aliases[0] = NULL;
	pvt->netent.n_net = tmp;

	return (&pvt->netent);
}

static struct netent *
nw_to_net(struct nwent *nwent) {
	struct pvt *pvt;
	u_long addr = 0;
	int i; 
	int msbyte;

	if (!nwent || nwent->n_addrtype != AF_INET)
		return (NULL);
	freepvt();
	net_data.nw_data = malloc(sizeof(struct pvt));
	if (!net_data.nw_data) {
		errno = ENOMEM;
		h_errno = NETDB_INTERNAL;
		return (NULL);
	}
	pvt = net_data.nw_data;
	pvt->netent.n_name = nwent->n_name;
	pvt->netent.n_aliases = nwent->n_aliases;
	pvt->netent.n_addrtype = nwent->n_addrtype;
	
/*
 * What this code does: Converts net addresses from network to host form.
 *                    
 * msbyte: the index of the most significant byte in the n_addr array.
 *
 * Shift bytes in significant order into addr. When all signicant
 * bytes are in, zero out bits in the LSB that are not part of the network.
 */
	msbyte = nwent->n_length / 8 +
		((nwent->n_length % 8) != 0 ? 1 : 0) - 1;
	for (i = 0; i <= msbyte; i++)
		addr = (addr << 8) | ((unsigned char *)nwent->n_addr)[i];
	i = (32 - nwent->n_length) % 8;
	if (i != 0)
		addr &= ~((1 << (i + 1)) - 1);
	pvt->netent.n_net = addr;
	return (&pvt->netent);
}


